cudax/include/cuda/experimental/__utility/basic_any/basic_any_ref.cuh

```diff
@@ -0,0 +1,357 @@
+///
+/// Note: a `basic_any<__ireference<_Interface>> &&` is an rvalue reference,
+/// whereas a `basic_any<_Interface &> &&` is an lvalue reference.
+template <class _Interface>
+struct __ireference : _Interface
+{
+  static_assert(_CUDA_VSTD::is_class_v<_Interface>, "expected a class type");
+  static constexpr size_t __size_      = sizeof(void*);
+  static constexpr size_t __align_     = alignof(void*);
+  static constexpr bool __is_const_ref = _CUDA_VSTD::is_const_v<_Interface>;
+
+  using interface _CCCL_NODEBUG_ALIAS = _CUDA_VSTD::remove_const_t<_Interface>;
+};
+
+#if !defined(__cpp_concepts)
+///
+/// A base class for basic_any<__ireference<_Interface>> that provides a
+/// conversion to basic_any<__ireference<_Interface const>>. Only used
+/// when concepts are not available.
+///
+template <class _Interface>
+struct __basic_any_reference_conversion_base
+{
+  _CCCL_NODISCARD _CUDAX_HOST_API operator basic_any<__ireference<_Interface const>>() const noexcept
+  {
+    return basic_any<__ireference<_Interface const>>(static_cast<basic_any<__ireference<_Interface>> const&>(*this));
+  }
+};
+
+template <class _Interface>
+struct __basic_any_reference_conversion_base<_Interface const>
+{};
+#endif
+
+_CCCL_DIAG_PUSH
+// "operator basic_any<...> will not be called for implicit or explicit conversions"
+_CCCL_DIAG_SUPPRESS_NVHPC(conversion_function_not_usable)
+// "operator basic_any<...> will not be called for implicit or explicit conversions"
+_CCCL_NV_DIAG_SUPPRESS(554)
+
+///
+/// \c basic_any<__ireference<_Interface>>
+///
+/// A `basic_any<__ireference<_Interface>>` is a reference to an object that
+/// satisfies the given interface. It is used as the result of dereferencing
+/// a `basic_any<_Interface*>` and as the implementation of
+/// `basic_any<_Interface&>`.
+///
+/// `basic_any<__ireference<_Interface>>` is neither copyable nor movable. It is
+/// not an end-user type.
+template <class _Interface, class Select>
+struct _LIBCUDACXX_DECLSPEC_EMPTY_BASES basic_any<__ireference<_Interface>, Select>
+    : __interface_of<__ireference<_Interface>>
+#if !defined(__cpp_concepts)
+    , __basic_any_reference_conversion_base<_Interface>
+#endif
+{
+  static_assert(_CUDA_VSTD::is_class_v<_Interface>, "expecting a class type");
+  using interface_type                 = _CUDA_VSTD::remove_const_t<_Interface>;
+  static constexpr bool __is_const_ref = _CUDA_VSTD::is_const_v<_Interface>;
+
+  basic_any(basic_any&&)      = delete;
+  basic_any(basic_any const&) = delete;
+
+  basic_any& operator=(basic_any&&)      = delete;
+  basic_any& operator=(basic_any const&) = delete;
+
+#if defined(__cpp_concepts)
+  /// \brief A non-const basic_any reference can be implicitly converted to a
+  /// const basic_any reference.
+  _CCCL_NODISCARD _CUDAX_HOST_API operator basic_any<__ireference<_Interface const>>() const noexcept
+    requires(!__is_const_ref)
+  {
+    return basic_any<__ireference<_Interface const>>(*this);
+  }
```

Comment:
We could make this work with concept emulation too
```suggestion
  /// \brief A non-const basic_any reference can be implicitly converted to a
  /// const basic_any reference.
  _CCCL_TEMPLATE(class _Interface2 = _Interface)
   _CCCL_REQUIRES((!_CUDA_VSTD::is_const_v<_Interface2>))
  _CCCL_NODISCARD _CUDAX_HOST_API operator basic_any<__ireference<_Interface const>>() const noexcept
  {
    return basic_any<__ireference<_Interface const>>(*this);
  }
```

Reply:
it's important that this conversion operation is a non-template. it is needed to make conversions happen at the right time.